### PR TITLE
Nginx Plus stats

### PR DESCRIPTION
**Pull request by mplonka**
Retrieves Nginx Plus stats from "stats" module.

### CODE REVIEW

**Comment by phemmer** [comment]
Just my own $0.02: The vast majority of the code seems to be mapping of data structures. What fields to look for in the nginx output, and then what fields to pass on, and how to name them.
I would encourage a more programmatic approach. I've been bitten by several plugins in telegraf that use hard-coded field mappings because either a new version of the software was released and telegraf hadn't been updated to support the fields, or the plugin author didn't think the fields were useful.

Also what about merging some of the measurements together, such as `nginx.requests`, `nginx.ssl`, `nginx.connections` and `nginx.processes`. This way were not overwhelming the user with a bunch of micro-measurements. For some of them they should be kept separate, because there are multiple series (tag permutations), but for the 4 mentioned, there's only one series in all of them.

But again, this is just my personal thoughts. I'm not a maintainer, so they should not be construed as requirements. :-)

**Comment by mplonka** [comment]
Thanks for your comments, PHemmer. I'm happy to modify the code to make it more consistent with other plugins and simplify this data mapping there. Is there any reference plugin that I should use as a reference?

**Comment by poblahblahblah** [comment]
Hello,

We would like to see nginx+ monitoring in telegraf. What needs to be done to get this revived and merged in? Is there a good reference plugin I can look at to address the comments @phemmer had? Do the telegraf maintainers preferences mirror those comments?

File: plugins/inputs/nginx/nginxplus.go [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+	s.gatherRequestMetrics(tags, acc)
+	s.gatherZoneMetrics(tags, acc)
+	s.gatherUpstreamMetrics(tags, acc)
+	s.gatherCacheMetrics(tags, acc)
+	s.gatherStreamMetrics(tags, acc)
+}
+
+func (s *Status) gatherProcessesMetrics(tags map[string]string, acc telegraf.Accumulator) {
+	acc.AddFields(
+		"nginx.processes",
```
Comment (danielnelson) [inline plugins/inputs/nginx/nginxplus.go]: Anywhere we are using dotted notation should either use snake_case or in some cases the subitems should become tags.  I think in this PR all should become snake_case.

This applies to measurement names and field names throughout the pull request.

File: plugins/inputs/nginx/nginxplus.go [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+
+Subsequent versions of status response structure available here:
+1. http://web.archive.org/web/20130805111222/http://nginx.org/en/docs/http/ngx_http_status_module.html
+2. http://web.archive.org/web/20131218101504/http://nginx.org/en/docs/http/ngx_http_status_module.html
+3. not available
+4. http://web.archive.org/web/20141218170938/http://nginx.org/en/docs/http/ngx_http_status_module.html
+5. http://web.archive.org/web/20150414043916/http://nginx.org/en/docs/http/ngx_http_status_module.html
+6. http://web.archive.org/web/20150918163811/http://nginx.org/en/docs/http/ngx_http_status_module.html
+7. http://web.archive.org/web/20161107221028/http://nginx.org/en/docs/http/ngx_http_status_module.html
+*/
```
Comment (danielnelson) [inline plugins/inputs/nginx/nginxplus.go]: This comment would go into the new nginxplus README

**Comment by mplonka** [comment]
Hi @danielnelson. Thanks for the review. Please let me know if this one looks better.

File: plugins/inputs/nginx/nginxplus.go [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+			Processing  int `json:"processing"`
+			Connections int `json:"connections"`
+			Sessions    *struct {
+				Total       int64 `json:"total"`
+				Sessions1xx int64 `json:"1xx"`
+				Sessions2xx int64 `json:"2xx"`
+				Sessions3xx int64 `json:"3xx"`
+				Sessions4xx int64 `json:"4xx"`
+				Sessions5xx int64 `json:"5xx"`
+			} `json:"sessions"`
```
Comment (danielnelson) [inline plugins/inputs/nginx/nginxplus.go]: I think this can be a HttpResponsesStats

File: plugins/inputs/nginx/nginxplus.go [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+		HandshakesFailed int64 `json:"handshakes_failed"`
+		SessionReuses    int64 `json:"session_reuses"`
+	} `json:"ssl"`
+
+	Requests struct {
+		Total   int64 `json:"total"`
+		Current int   `json:"current"`
+	} `json:"requests"`
+
+	ServerZones map[string]struct { // added in version 2
```
Comment (danielnelson) [inline plugins/inputs/nginx/nginxplus.go]: Does this need to be a pointer since it is not in version 1?

File: plugins/inputs/nginx/nginxplus.go [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+		Keepalive int       `json:"keepalive"`
+		Zombies   int       `json:"zombies"` // added in version 6
+		Queue     *struct { // added in version 6
+			Size      int   `json:"size"`
+			MaxSize   int   `json:"max_size"`
+			Overflows int64 `json:"overflows"`
+		} `json:"queue"`
+	} `json:"upstreams"`
+
+	Caches map[string]struct { // added in version 2
```
Comment (danielnelson) [inline plugins/inputs/nginx/nginxplus.go]: Does this need to be a pointer since it is not in version 1?

File: plugins/inputs/nginx/README.md [hunk 6]
```diff
@@ -45,3 +45,24 @@ It produces:
...
+
+- version 3 - not available
+
+- [version 4](http://web.archive.org/web/20141218170938/http://nginx.org/en/docs/http/ngx_http_status_module.html)
+
+- [version 5](http://web.archive.org/web/20150414043916/http://nginx.org/en/docs/http/ngx_http_status_module.html)
+
+- [version 6](http://web.archive.org/web/20150918163811/http://nginx.org/en/docs/http/ngx_http_status_module.html)
+
+- [version 7](http://web.archive.org/web/20161107221028/http://nginx.org/en/docs/http/ngx_http_status_module.html)
```
Comment (danielnelson) [inline plugins/inputs/nginx/README.md]: How does someone know what version they have?  Does this plugin support all versions?

Comment (danielnelson) [inline plugins/inputs/nginx/README.md]: We can probably cover all the version by linking to https://nginx.org/en/docs/http/ngx_http_status_module.html#compatibility

File: plugins/inputs/nginx/nginxplus.go (hunk 4 above)
Comment (mplonka) [inline plugins/inputs/nginx/nginxplus.go]: It's completely absent from version 1

File: plugins/inputs/nginx/nginxplus.go (hunk 5 above)
Comment (mplonka) [inline plugins/inputs/nginx/nginxplus.go]: Same as above: absent from version 1

File: plugins/inputs/nginx/README.md (hunk 6 above)
Comment (mplonka) [inline plugins/inputs/nginx/README.md]: The plugin supports all versions. Version is sent in the very first attribute.

File: plugins/inputs/nginx/nginxplus.go (hunk 3 above)
Comment (mplonka) [inline plugins/inputs/nginx/nginxplus.go]: 100%
Renaming this struct to make the name still meaningful in this context

**Comment by mplonka** [comment]
I do not have strong views on adding Nginx Plus support to this plugin or creating separate one. Keeping it here seems more intuitive for Nginx Plus users though.

**Comment by danielnelson** [comment]
Please make it a new plugin `nginxplus`.  The reason for this is that it has unique setup and format from the regular nginx so there is little to gain from combining them.

**Comment by poblahblahblah** [comment]
Hey @mplonka - can I help in any way get this plugin separated out into a separate plugin?

**Comment by poblahblahblah** [comment]
Hello,

Is there anything I can do to get this separated out into a separate plugin?

**Comment by danielnelson** [comment]
@poblahblahblah Could you fork @mplonka's code and update the plugin per the review?

**Comment by poblahblahblah** [comment]
@danielnelson Would you like these two plugins to be completely separate? It looks like the current way it's written the nginxplus plugin use a number of functions from the nginx plugin. I can separate these completely or I can just import the existing nginx plugin into nginxplus.

Any guidance would be appreciated :)

**Comment by danielnelson** [comment]
I think they should probably be completely independent.  I didn't see much shared code, is there something specific you are looking at?

**Comment by danielnelson** [comment]
Added in #3214

**Pull request closed without merging**